### PR TITLE
Feature/28 dense metric

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

Refactor of function `sample` and object `SamplerArgs`.

 + Better argument names - as similar as possible to RStan and/or CmdStan
 + Full set of options for all sampler arguments
 + Proper documentation for all sampler arguments

Previously, the `sample` function had too many arguments, e.g., because input data and inits can be either a Python dictionary or a filepath string, there were pairs of arguments `data` and `data_file`, etc.   Changed to single argument which allows specified union of types.

The `sample` function hadn't implemented the full set of CmdStan options - in particular:

 + no option of using a dense mass matrix for the metric
 + no option for per-chain seed, step-size, metric
 + no option for user-specified chain ids
 + no option for specifying warmup schedule (init_buffer, window, term_buffer)

Now the `metric` argument takes either the string `diag` or `dense`, or
a filename or list of filenames.  Added utility functions to scan JSON or Rdump
file to get dimensions of the `inv_metric`

Instead of exposing argument `init_buffer`, `window`, and `term_buffer`
the `sample` command takes a tuple `warmup_schedule` which specifies
the % of iterations allocated to each phase of warmup.  This is in line with
the warning message that the sampler returns when there are fewer than
the sum of the default values for `init_buffer`, `window`, and `term_buffer`.
This avoids the confusing issue of what the `window` argument means
(initial size of window which keeps doubling during adaptation of the mass matrix).

The `SamplerArgs` `validate` function now validates all arguments and
checks inter-arg dependencies as much as possible.  There are unit tests
for good/bad configurations.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):  Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: cmdstanpy/cmds.py [hunk 1]
```diff
@@ -12,32 +12,34 @@
...
 from cmdstanpy import TMPDIR
 from cmdstanpy.lib import Model, StanData, RunSet, SamplerArgs
 from cmdstanpy.utils import cmdstan_path
 
 
 def compile_model(
-    stan_file: str = None, opt_lvl: int = 1, overwrite: bool = False,
-    include_paths: List[str] = None
+    stan_file: str = None,
+    opt_lvl: int = 1,
```
Comment (ahartikainen) [inline cmdstanpy/cmds.py]: should the default be 3?

File: cmdstanpy/cmds.py [hunk 2]
```diff
@@ -12,32 +12,34 @@
...
+    include_paths: List[str] = None,
 ) -> Model:
     """
     Compile the given Stan model file to an executable.
 
     :param stan_file: Path to Stan program
 
-    :param opt_lvl: Optimization level for c++ compiler, one of {0, 1, 2, 3}
+    :param opt_lvl: Optimization level for C++ compiler, one of {0, 1, 2, 3}
       where level 0 optimization results in the shortest compilation time
```
Comment (ahartikainen) [inline cmdstanpy/cmds.py]: On Windows at least with PyStan my python crashes with `-O0`

File: cmdstanpy/cmds.py [hunk 3]
```diff
@@ -79,140 +81,226 @@ def compile_model(
...
+        as a dictionary with entries matching the data variables,
+        or as the path of a data file in JSON or Rdump format.
+
+    :param chains: Number of sampler chains, should be > 1.
+
+    :param cores: Number of processes to run in parallel. If this value
+        exceeds the number of available processors, only max processors chains
+        will be run in parallel.
+
+    :param seed: The seed for random number generator or a list of per-chain
```
Comment (ahartikainen) [inline cmdstanpy/cmds.py]: What is the maximum value accepted?

File: cmdstanpy/cmds.py [hunk 4]
```diff
@@ -79,140 +81,226 @@ def compile_model(
...
+        only the diagonal elements.
+
+        The metric is initialized to a unit matrix.  If the value of the
+        metric argument is a string other than `diag` or `dense` or a
+        list of strings, these are treated as pathnames to a data file
+        in JSON or Rdump format which contains entry `inv_metric` which
+        is either a vector or a full matrix, depending on whether the
+        diagonal or dense covariance matrix is to be estimated.
+
+        The length of the list of pathnames must match the number of chains.
```
Comment (ahartikainen) [inline cmdstanpy/cmds.py]: pathnames -> paths (?)

Comment (ahartikainen) [reply cmdstanpy/cmds.py]: Also, what happens if the path is duplicated? On Windows that will probably crash, unless each file is copied to unique location (tempfile)

File: cmdstanpy/lib.py [hunk 5]
```diff
@@ -81,194 +82,360 @@ def data_file(self) -> str:
...
+        if self.metric is not None:
+            dims = None
+            if isinstance(self.metric, str):
+                if self.metric == 'diag':
+                    self.metric = 'diag_e'
+                elif self.metric == 'dense':
+                    self.metric = 'dense_e'
+                else:
+                    if not os.path.exists(self.metric):
+                        raise ValueError('no such file {}'.format(self.metric))
```
Comment (ahartikainen) [inline cmdstanpy/lib.py]: I think self.metric could also accept `diag_e` and `dense_e` kwargs

File: cmdstanpy/cmds.py (hunk 3 above)
Comment (mitzimorris) [reply cmdstanpy/cmds.py]: 2^32-1  - will doc

File: cmdstanpy/cmds.py (hunk 4 above)
Comment (mitzimorris) [reply cmdstanpy/cmds.py]: no checks on pathname uniqueness - can add.

File: cmdstanpy/lib.py (hunk 5 above)
Comment (mitzimorris) [reply cmdstanpy/lib.py]: good point - will do.

File: cmdstanpy/cmds.py (hunk 1 above)
Comment (mitzimorris) [reply cmdstanpy/cmds.py]: that slows down compilation time which is already quite slow.  once we get compilation sped up (Bob's working on that now), we can change default value.

File: cmdstanpy/lib.py (hunk 5 above)
Comment (mitzimorris) [reply cmdstanpy/lib.py]: the same goes for the inits arg as well, no?  checking there as well.

File: cmdstanpy/cmds.py (hunk 4 above)
Comment (mitzimorris) [reply cmdstanpy/cmds.py]: list of `inits` arg paths must be unique as well, no?  adding checks on both args.

File: cmdstanpy/cmds.py (hunk 1 above)
Comment (ahartikainen) [reply cmdstanpy/cmds.py]: Is there a difference in sampling time?

Comment (mitzimorris) [reply cmdstanpy/cmds.py]: a year ago there was some discussion on discourse:  https://discourse.mc-stan.org/t/o2-vs-o3-compiler-optimization-level/3714

we could use boolean arg `optimize` where False is optimization level 1 and True is optimization level either 2 or 3 (sounds like 2 might be preferable)

**Pull request merged**